doc/PROGMEM.rst

```diff
@@ -0,0 +1,266 @@
+and manipulate flash strings is not straight forward as the esp8266 must
+read from flash in 4byte words. In the Arduino IDE for esp8266 there are
+several functions that can help retrieve strings from flash that have
+been stored using PROGMEM. Both of the examples above will give you a
+``const char *`` back, however if you try to do anything with these
+pointers, without correct 32bit alignment you will get a seg fault and
+the esp will crash. You must read from the flash 32 bit aligned.
+
+Functions to read back from PROGMEM
+-----------------------------------
+
+Which are all defined in
+`pgmspace.h <https://github.com/esp8266/Arduino/blob/master/cores/esp8266/pgmspace.h>`__
+
+.. code:: cpp
+
+    int memcmp_P(const void* buf1, PGM_VOID_P buf2P, size_t size);
+    void* memccpy_P(void* dest, PGM_VOID_P src, int c, size_t count);
+    void* memmem_P(const void* buf, size_t bufSize, PGM_VOID_P findP, size_t findPSize);
+    void* memcpy_P(void* dest, PGM_VOID_P src, size_t count);
+    char* strncpy_P(char* dest, PGM_P src, size_t size);
+    #define strcpy_P(dest, src)          strncpy_P((dest), (src), SIZE_IRRELEVANT)
+    char* strncat_P(char* dest, PGM_P src, size_t size);
+    #define strcat_P(dest, src)          strncat_P((dest), (src), SIZE_IRRELEVANT)
+    int strncmp_P(const char* str1, PGM_P str2P, size_t size);
+    #define strcmp_P(str1, str2P)          strncmp_P((str1), (str2P), SIZE_IRRELEVANT)s
+    int strncasecmp_P(const char* str1, PGM_P str2P, size_t size);
+    #define strcasecmp_P(str1, str2P)          strncasecmp_P((str1), (str2P), SIZE_IRRELEVANT)
+    size_t strnlen_P(PGM_P s, size_t size);
+    #define strlen_P(strP)          strnlen_P((strP), SIZE_IRRELEVANT)
+    char* strstr_P(const char* haystack, PGM_P needle);
+    int printf_P(PGM_P formatP, ...);
+    int sprintf_P(char *str, PGM_P formatP, ...);
+    int snprintf_P(char *str, size_t strSize, PGM_P formatP, ...);
+    int vsnprintf_P(char *str, size_t strSize, PGM_P formatP, va_list ap);
+
+There are a lot of functions there but in reality they are ``_P``
+versions of standard c functions that are adapted to read from the
+esp8266 32bit aligned flash. All of them take a ``PGM_P`` which is
+essentially a ``const char *``. Under the hood these functions all use:
+
+.. code:: cpp
+
+    #define pgm_read_byte(addr)                                                    \
+    (__extension__({                                                               \
+        PGM_P __local = (PGM_P)(addr);  /* isolate varible for macro expansion */         \
+        ptrdiff_t __offset = ((uint32_t)__local & 0x00000003); /* byte aligned mask */            \
+        const uint32_t* __addr32 = (const uint32_t*)((const uint8_t*)(__local)-__offset); \
+        uint8_t __result = ((*__addr32) >> (__offset * 8));                        \
+        __result;                                                                  \
+    }))
```

Comment:
This bit of code is no longer used, because the compiler used to optimize it away in some cases. Suggest adding a link to the pgmspace.h header instead.